### PR TITLE
expression: fix the issue of incorrect result when handling unary minus for time

### LINKED ISSUE #52262: Unexpected values appeared in select query statements that should logically imply each other
## Bug Report

Please answer these questions before submitting your issue. Thanks!

 In theory,the result of sql2(distinct)  ∈  the result of sql1

 however,the value -20171112000000.100000 change to -2017111200000 after adding DISTINCT,which seems like a logical bug
### 1. Minimal reproduce step (Required)
```
create table table_3_utf8_undef (
`pk` int primary key,
`col_bigint_undef_signed` bigint  ,
`col_bigint_undef_unsigned` bigint unsigned ,
`col_bigint_key_signed` bigint  ,
`col_bigint_key_unsigned` bigint unsigned ,
`col_float_undef_signed` float  ,
`col_float_undef_unsigned` float unsigned ,
`col_float_key_signed` float  ,
`col_float_key_unsigned` float unsigned ,
`col_double_undef_signed` double  ,
`col_double_undef_unsigned` double unsigned ,
`col_double_key_signed` double  ,
`col_double_key_unsigned` double unsigned ,
`col_decimal(40, 20)_undef_signed` decimal(40, 20)  ,
`col_decimal(40, 20)_undef_unsigned` decimal(40, 20) unsigned ,
`col_decimal(40, 20)_key_signed` decimal(40, 20)  ,
`col_decimal(40, 20)_key_unsigned` decimal(40, 20) unsigned ,
`col_char(20)_undef_signed` char(20)  ,
`col_char(20)_key_signed` char(20)  ,
`col_varchar(20)_undef_signed` varchar(20)  ,
`col_varchar(20)_key_signed` varchar(20)  ,
key (`col_bigint_key_signed`),
key (`col_bigint_key_unsigned`),
key (`col_float_key_signed`),
key (`col_float_key_unsigned`),
key (`col_double_key_signed`),
key (`col_double_key_unsigned`),
key (`col_decimal(40, 20)_key_signed`),
key (`col_decimal(40, 20)_key_unsigned`),
key (`col_char(20)_key_signed`),
key (`col_varchar(20)_key_signed`)
) character set utf8 ;
insert into table_3_utf8_undef values (0,82.1847,1,39.0425,38.1089,-1,1,94.1106,1.009,12.991,19755,-13064,0,1,79.1429,-2,1,'well','3
','-0','e'),(1,1,20.0078,-9.183,68.1957,1,2,1,0.0001,12.991,2,71.0510,1,-1,2,12.991,12.991,'3	','1','3	','-0'),(2,-2,1,-21247,1.009,2,1.009,0.0001,36.0002,-2,2,-0,0.0001,-2,0.1598,47.1515,1.009,'3	','w','-1','e');
create table table_7_utf8_undef (
`pk` int primary key,
`col_bigint_undef_signed` bigint  ,
`col_bigint_undef_unsigned` bigint unsigned ,
`col_bigint_key_signed` bigint  ,
`col_bigint_key_unsigned` bigint unsigned ,
`col_float_undef_signed` float  ,
`col_float_undef_unsigned` float unsigned ,
`col_float_key_signed` float  ,
`col_float_key_unsigned` float unsigned ,
`col_double_undef_signed` double  ,
`col_double_undef_unsigned` double unsigned ,
`col_double_key_signed` double  ,
`col_double_key_unsigned` double unsigned ,
`col_decimal(40, 20)_undef_signed` decimal(40, 20)  ,
`col_decimal(40, 20)_undef_unsigned` decimal(40, 20) unsigned ,
`col_decimal(40, 20)_key_signed` decimal(40, 20)  ,
`col_decimal(40, 20)_key_unsigned` decimal(40, 20) unsigned ,
`col_char(20)_undef_signed` char(20)  ,
`col_char(20)_key_signed` char(20)  ,
`col_varchar(20)_undef_signed` varchar(20)  ,
`col_varchar(20)_key_signed` varchar(20)  ,
key (`col_bigint_key_signed`),
key (`col_bigint_key_unsigned`),
key (`col_float_key_signed`),
key (`col_float_key_unsigned`),
key (`col_double_key_signed`),
key (`col_double_key_unsigned`),
key (`col_decimal(40, 20)_key_signed`),
key (`col_decimal(40, 20)_key_unsigned`),
key (`col_char(20)_key_signed`),
key (`col_varchar(20)_key_signed`)
) character set utf8 ;
insert into table_7_utf8_undef values (0,-9.183,1,1.1384,2,15.1271,12.991,-2,0.0001,36.1270,79.1819,0.0001,0.0001,3.1387,52.0818,-0,0.0001,'1','3	','0','0'),(1,79,12.991,107,2,-0.0001,0,1.009,1.009,34,1,-1,69.0208,1,2,120,12.991,'3	','-1','if','b'),(2,-2,1,-9.183,1,12.991,0.0001,53,12.991,1.009,12.991,12.991,0.0001,-0.0001,12.991,0.0001,2,'3
','p','0','3	'),(3,-0.0001,12.991,1.009,1.009,-9.183,2,0,1,-2,1,2,1,2,1.009,2,12.991,'3
','0','k','0'),(4,1.009,0.0001,-1,12.991,2,47,2,0,12.991,12.991,1.009,0,1.009,1.009,-0.0001,6949,'-1','	3','1','m'),(5,-0,1,0,0,0.0001,28.1237,12.991,0,12.991,12.991,-0,12.991,2,2,2,1.009,'0','	3','0','	3'),(6,45.0855,1,38.1166,1,1.009,80.0284,2,122,0.0001,0,-1,11130,0,1,1,0,'know','-0','
3','3
');
```
sql1
```
SELECT 
  (_UTF8MB4'r' - OCTET_LENGTH(`f5`) * SQRT(8)) AS `f2`,
  (-`f4`) AS `f3`
FROM 
  (SELECT `col_decimal(40, 20)_key_signed` AS `f5` 
   FROM `table_3_utf8_undef`) AS `t1`
NATURAL JOIN 
  (SELECT (DATE_ADD(DATE(_UTF8MB4'2017-11-12 08:48:25'), INTERVAL 1 HOUR_MICROSECOND)) AS `f4` 
   FROM 
     (SELECT `col_double_undef_unsigned` AS `f10`,
             `col_decimal(40, 20)_undef_unsigned` AS `f12` 
      FROM `table_7_utf8_undef`) AS `t2` 
   WHERE 
     ((((COERCIBILITY(`f12`)) BETWEEN 1 AND _UTF8MB4'l') IS FALSE))) AS `t3`;

```
sql2
```
SELECT DISTINCT
  (_UTF8MB4'r' - OCTET_LENGTH(`f5`) * SQRT(8)) AS `f2`,
  (-`f4`) AS `f3`
FROM
  (SELECT `col_decimal(40, 20)_key_signed` AS `f5`
   FROM `table_3_utf8_undef`) AS `t1`
NATURAL JOIN
  (SELECT (DATE_ADD(DATE(_UTF8MB4'2017-11-12 08:48:25'), INTERVAL 1 HOUR_MICROSECOND)) AS `f4`
   FROM
     (SELECT `col_double_undef_unsigned` AS `f10`,
             `col_decimal(40, 20)_undef_unsigned` AS `f12`
      FROM `table_7_utf8_undef`) AS `t2`
   WHERE
     ((((COERCIBILITY(`f12`)) BETWEEN 1 AND _UTF8MB4'l') IS FALSE))) AS `t3`;

```
### 2. What did you expect to see? (Required)
sql1
```
 f2         |           f3           
--------------------+------------------------
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
 -65.05382386916237 | -20171112000000.100000 
```
sql2
```
 f2         |           f3           
--------------------+------------------------
 -65.05382386916237 | -20171112000000.100000 
```

### 3. What did you see instead (Required)
sql2
```
     f2         |       f3        
--------------------+-----------------
 -65.05382386916237 | -20171112000000 
(1 row)
```

### 4. What is your TiDB version? (Required)
tidb 7.6

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: close #52262

Problem Summary:

- `func (c *unaryMinusFunctionClass) getFunction` doesn't set decimal represent info for time type expression.

### What changed and how does it work?

- Cast time type expression into decimal in `func (c *unaryMinusFunctionClass) getFunction` .

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue of incorrect result when handling unary minus for time
```

### CODE REVIEW

File: pkg/expression/integration_test/integration_test.go [hunk 1]
```diff
@@ -2588,6 +2588,28 @@ func TestTimeBuiltin(t *testing.T) {
...
+	// date
+	result = tk.MustQuery(`select distinct -(DATE('2017-11-12 08:48:25'))`)
+	result.Check(testkit.Rows("-20171112"))
+	result = tk.MustQuery(`select distinct (DATE('2017-11-12 08:48:25'))`)
+	result.Check(testkit.Rows("2017-11-12"))
+	// timestamp
+	result = tk.MustQuery(`select distinct (Timestamp('2017-11-12 08:48:25'))`)
+	result.Check(testkit.Rows("2017-11-12 08:48:25"))
+	result = tk.MustQuery(`select distinct -(Timestamp('2017-11-12 08:48:25'))`)
+	result.Check(testkit.Rows("-20171112084825"))
```
Comment (XuHuaiyu) [inline pkg/expression/integration_test/integration_test.go]: All the input arguments are constant.
It'll be better to add some test cases, which use columns as input arguments.

Comment (solotzg) [reply pkg/expression/integration_test/integration_test.go]: done

**Review by XuHuaiyu — APPROVED** [review]

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**